python_modules/dagster/dagster/_core/asset_graph_view/asset_graph_view.py

```diff
@@ -146,6 +153,73 @@ def only_partition_keys(self, partition_keys: AbstractSet[PartitionKey]) -> "Ass
             & self._compatible_subset,
         )
 
+    @cached_property
+    def latest_time_window_slice(self) -> "AssetSlice":
+        """Returns the latest time window for the asset slice.
+
+        If the asset slice is time-window-partitiond (e.g. daily, hourly),
+        this will return a slice for latest complete time window,
+        relative to the current effective time.
+
+        If the asset slice is not time-windowed, this will return the same asset slice.
+
+        TODO: add language for multi-dimensional partitioning when we support it
+        TODO: add language for dynamic partitioning when we support it
+        """
+        if self._partitions_def is None:
+            return self
+
+        if isinstance(self._partitions_def, StaticPartitionsDefinition):
+            return self
+
+        if isinstance(self._partitions_def, TimeWindowPartitionsDefinition):
+            time_window = self._partitions_def.get_last_partition_window(
+                self._asset_graph_view.effective_dt
+            )
+
+            return (
+                self._asset_graph_view.create_from_time_window(self.asset_key, time_window)
+                if time_window
+                else self._asset_graph_view.create_empty_slice(self.asset_key)
+            )
+
+        # Need to handle dynamic and multi-dimensional partitioning
+        check.failed(f"Unsupported partitions_def: {self._partitions_def}")
+
+    @cached_property
+    def latest_time_window(self) -> TimeWindow:
+        """If the underlying asset is time-window partitioned, this will return the latest complete
+        time window relative to the effective date. For example if it is daily partitioned starting
+        at midnight every day.  If the effective date is before the start of the partition definition, this will
+        return the empty time window (where both start and end are datetime.max).
+
+        If the underlying asset is unpartitioned or static partitioned and it is not empty,
+        this will return a time window from the beginning of time to the effective date. If
+        it is empty it will return the empty time window.
```

Comment:
>this will return the latest complete time window relative to the effective date

I'm confused as to why this is a method on `AssetSlice` if it operates with respect to the underlying `AssetGraphView` upper temporal bound rather than the upper temporal bound of the asset slice itself (which would have to come from the contained `_AssetSliceCompatibleSubset` right?).

Reply:
> Not sure if this is consequential but it doesn't make sense to me why the time window associated with unpartitioned/static-partitioned assets should depend on whether the slice is empty. IMO the time window should always be "from the beginning of time to the effective date".

This makes sense to me.

Reply:
> I'm confused as to why this is a method on AssetSlice if it operates with respect to the underlying AssetGraphView upper temporal bound rather than the upper temporal bound of the asset slice itself (which would have to come from the contained _AssetSliceCompatibleSubset right?).

They operate against the same temporal bound. `AssetSlice` has a reference to `AssetGraphView`. It is a method on `AssetSlice` for pure convenience purposes.

Reply:
>They operate against the same temporal bound. AssetSlice has a reference to AssetGraphView. It is a method on AssetSlice for pure convenience purposes.

Isn't `AssetSlice` an arbitrary subset of asset partitions against the container set defined by the associated `AssetGraphView` temporal bound?

i.e. If I have a daily partitioned asset, and the `AssetGraphView` has a temporal bound of 2023/12/31, can't I have an `AssetSlice` that covers e.g. dates 2023/12/1 - 2023/12/7? For this slice, I would expect `AssetSlice.latest_time_window` to return the time window corresponding to 2023/12/7.